### PR TITLE
LibGL+LibGPU+LibSoftGPU: Implement ALL* the pixel formats

### PR DESCRIPTION
**Pull request by GMTA**
<sup>* Note: please check your local dealership to find out whether your pixel format is eligible to be included in "ALL"</sup>

A GPU (driver) is now responsible for reading and writing pixels from and to user data. The client (LibGL) is responsible for specifying how the user data must be interpreted or written to.

This allows us to centralize all pixel format conversion in one class, `LibSoftGPU::PixelConverter`. For both the input and output image, it takes a specification containing the image dimensions, the pixel type and the selection (basically a clipping rect), and converts the pixels from the input image to the output image.

Additionally, this adds support for the requested internal texel format for textures.

Effectively this means we now support almost all OpenGL 1.5 formats, and all custom logic has disappeared from:
  - `glDrawPixels`
  - `glReadPixels`
  - `glTexImage2D`
  - `glTexSubImage2D`

The new logic is still unoptimized, but on my machine I experienced no noticeable slowdown. :^)

ScummVM + Monkey4 before:

image

ScummVM + Monkey4 after:

image

### CODE REVIEW

File: Userland/Libraries/LibSoftGPU/PixelConverter.cpp [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+        return read_components.template operator()<u16, u16>();
+    default:
+        VERIFY_NOT_REACHED();
+    }
+}
+
+static constexpr void write_pixel_data(u8** output_data, float value, GPU::ImageDataLayout layout)
+{
+    auto write_value = [&output_data, &layout]<typename T>(T value) -> void {
+        if (layout.pixel_type.bits == GPU::PixelComponentBits::AllBits) {
```
Comment (GMTA) [inline Userland/Libraries/LibSoftGPU/PixelConverter.cpp:229]: Note to self: remove this and the `layout` argument; `write_pixel_data` is only invoked for `::AllBits`

File: Userland/Libraries/LibGL/Texture.cpp [hunk 2]
```diff
@@ -306,26 +308,20 @@ void GLContext::gl_tex_gen_floatv(GLenum coord, GLenum pname, GLfloat const* par
     m_texcoord_generation_dirty = true;
 }
 
+// FIXME: talk to GPU::Device to determine supported GPU::PixelTypes
+constexpr GPU::PixelType rasterizer_fixed_pixel_type = {
```
Comment (GMTA) [inline Userland/Libraries/LibGL/Texture.cpp:312]: Note to self: rename to something texture-related

File: Userland/Libraries/LibGPU/ImageFormat.h [hunk 3]
```diff
@@ -1,43 +1,179 @@
...
+        return { 4, 4, 4, 4 };
+    case PixelComponentBits::B5_5_5_1:
+        return { 5, 5, 5, 1 };
+    case PixelComponentBits::B5_6_5:
+        return { 5, 6, 5 };
+    case PixelComponentBits::B8_8_8_8:
+        return { 8, 8, 8, 8 };
+    case PixelComponentBits::B10_10_10_2:
+        return { 10, 10, 10, 2 };
+    default:
```
Comment (GMTA) [inline Userland/Libraries/LibGPU/ImageFormat.h:140]: Note to self: remove all `default:` cases from enum switches to make use of compiler errors in the future

File: Userland/Libraries/LibSoftGPU/Device.cpp [hunk 4]
```diff
@@ -1451,25 +1533,12 @@ void Device::set_light_model_params(GPU::LightModelParameters const& lighting_mo
...
-}
-
-NonnullRefPtr<GPU::Image> Device::create_image(GPU::ImageFormat format, unsigned width, unsigned height, unsigned depth, unsigned levels, unsigned layers)
+NonnullRefPtr<GPU::Image> Device::create_image(GPU::PixelType const& pixel_type, u32 width, u32 height, u32 depth, u32 levels, u32 layers)
 {
-    VERIFY(format == GPU::ImageFormat::BGRA8888);
+    VERIFY(pixel_type.format == GPU::PixelFormat::RGBA
+        && pixel_type.bits == GPU::PixelComponentBits::AllBits
+        && pixel_type.data_type == GPU::PixelDataType::Float
+        && pixel_type.components_order == GPU::ComponentsOrder::Normal);
```
Comment (GMTA) [inline Userland/Libraries/LibSoftGPU/Device.cpp:1541]: Note to self: also check packing

**Comment by sunverwerth** [comment]
Big changes here but LGTM! Do you want to address all your "note to self" comments before merging? Otherwise I'd approve this.

**Comment by GMTA** [comment]
> Big changes here but LGTM! Do you want to address all your "note to self" comments before merging? Otherwise I'd approve this.

I'd like to address those, yeah - I'll give you a heads up when the PR is updated

File: AK/FloatingPoint.h [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+    // Deconstruct input bits to float components
+    u64 from_sign = (input.bits() >> (From::exponentbits + From::mantissabits)) & From::signbit;
+    u64 from_exponent = (input.bits() >> From::mantissabits) & ((1UL << From::exponentbits) - 1);
+    u64 from_mantissa = input.bits() & ((1UL << From::mantissabits) - 1);
+
+    u64 to_sign = from_sign & To::signbit;
+    u64 to_exponent;
+    u64 to_mantissa;
+    auto target_value = [&to_sign, &to_exponent, &to_mantissa]() constexpr -> auto {
```
Comment (GMTA) [inline AK/FloatingPoint.h:69]: Note to self: remove all these `constexpr`s from lambdas

**Review by sunverwerth — APPROVED** [review]
LGTM!

**Pull request merged**